hikari/impl/shard.py

```diff
@@ -519,61 +516,63 @@ def intents(self) -> intents_.Intents:
 
     @property
     def is_alive(self) -> bool:
-        return self._ws is not None and not self._ws.sent_close
+        return self._keep_alive_task is not None
+
+    @property
+    def is_connected(self) -> bool:
+        return self._ws is not None and self._handshake_event is not None and self._handshake_event.is_set()
 
     @property
     def shard_count(self) -> int:
         return self._shard_count
 
     async def close(self) -> None:
-        self._check_if_alive()
-        if not self._closing_event.is_set():
-            try:
-                if self._ws is not None:
-                    self._logger.debug(
-                        "shard.close() was called and the websocket was still alive -- "
-                        "disconnecting immediately with GOING AWAY"
-                    )
-                    await self._ws.send_close(code=errors.ShardCloseCode.GOING_AWAY, message=b"shard disconnecting")
-                self._closing_event.set()
-            finally:
-                self._chunking_rate_limit.close()
-                self._total_rate_limit.close()
+        if not self._keep_alive_task:
+            raise errors.ComponentStateConflictError("Cannot close an inactive shard")
+
+        if self._is_closing:
+            await self.join()
+            return
 
-        # Wait for the shard to fully close
-        await self._closed_event.wait()
+        self._logger.info("shard has been requested to shutdown")
+        self._is_closing = True
 
-    async def get_user_id(self) -> snowflakes.Snowflake:
-        await self._handshake_completed.wait()
-        if self._user_id is None:
-            raise RuntimeError("user_id was not known, this is probably a bug")
+        self._keep_alive_task.cancel()
+        try:
+            await self._keep_alive_task
+        except asyncio.CancelledError:
+            pass
+
+        self._keep_alive_task = None
+        self._non_priority_rate_limit.close()
+        self._total_rate_limit.close()
+        self._is_closing = False
+        self._logger.info("shard shutdown successfully")
+
+    def get_user_id(self) -> snowflakes.Snowflake:
```

Comment:
what's the point of this breaking change?

Reply:
With the new lifetime, we can ensure it will always be available, so there is no need for it to be async anymore

Reply:
>         If the shard has not connected fully yet, this should wait until the ID
>         is set before returning.


The old code was just waiting until the shard had started to return the user (and returning if the shard was already active); how are you ensuring that it's available before the shard has started?

Reply:
With the new lifetime, we can ensure that after `start` completes, we will have the information available and there wont be an attempt to access it until we return from `start`, which is when it would be reasonably expected for the information to be available. There is no need for the function to be async, as you shouldnt be waiting for the user_id if you never start the shard.

Reply:
> There is no need for the function to be async, as you shouldnt be waiting for the user_id if you never start the shard.

Ppl could call that method before the shard has started, this is why it was async before to wait for the shard to start. Idk what you're getting at here.